### ISSUE TITLE
Alternative instance for ContT

### ISSUE THREAD

**Issue by sjshuck** [haskell/transformers#77]
<p><b><i>[original <a href="https://hub.darcs.net/ross/transformers/issue/69">issue 69</a> by dfranke]</i></b></p>
<pre><code>instance (Monoid r, Applicative m) =&gt; Alternative (ContT r m) where
    empty = ContT $ const (pure mempty)
    (ContT f) &lt;|&gt; (ContT g) = ContT $ \c -&gt; liftA2 (&lt;&gt;) (f c) (g c)</code></pre>

<p>Some ways this can be useful:</p>

<ol style="list-style-type: decimal">
<li><p>You're using <code>ContT</code> to implement a backtracking search. <code>r</code> is <code>Maybe (First Result))</code> for some <code>Result :: *</code>. <code>a &lt;|&gt; b</code> then means "try <code>a</code> then try <code>b</code>".</p></li>
<li><p>You have a function <code>inferType :: Expr a -&gt; (a -&gt; Type) -&gt; Type</code> where <code>Expr a</code> represents an AST containing free variables in <code>a</code>, and <code>(a -&gt; Type)</code> is an environment which maps free variables to their type. <code>Type</code> has a monoid instance where <code>mempty</code> is the type of bottom and <code>(&lt;&gt;)</code> is unification; you form a <code>Cont Type a</code> by way of <code>cont (inferType e)</code>. Then <code>(&lt;|&gt;)</code> will lift <code>(&lt;&gt;)</code> from working in the empty environment to working in arbitrary ones.</p></li>
</ol>

**Comment by sjshuck** [comment]
Just want to keep in mind that `empty >>= f` is hinting at `MonadPlus` laws.

**Comment by sjshuck** [comment]
Proofs of concepts: https://gist.github.com/sjshuck/966caa9797acb073dfb3c755bbeed839

**Comment by sergv** [comment]
I think the slightly revised second proposal

```haskell
instance (Alternative m) => Alternative (ContT r m) where
    empty = ContT $ \c -> empty >>= c
    (ContT f) <|> (ContT g) = ContT $ \c -> f c <|> g c
```

wound still satisfy `empty >>= f = empty` by deferring to this law holding for the underlying `m`.

**Comment by sjshuck** [comment]
@sergv Yeah so that needs `Monad m` in the context and is maybe relevant to `MonadPlus` rather than `Alternative`.  Which is probably relevant to our ticket inasmuch as those two instances ought to be functionally equivalent.

**Comment by sjshuck** [comment]
@aspiwack
I notice you made this instance for your indexed continuation monad.  https://github.com/tweag/stacked/blob/f5ab1c8723ede3afada8d981851c8a5b513e8d27/src/Control/Monad/Indexed/Cont2.hs#L112-L114

Do you have an opinion about this issue?

**Comment by aspiwack** [comment]
@sjshuck Mine is a two-continuation monad (one success and one failure continuation), the alternative instance is obtained by passing the right-hand side as the failure continuation of the left-hand side. It's quite canonical too: if you take the pure two-continuation monad and quantify universally over the return type `r`, you obtain something which is equivalent to the list monad, and `<|>` is concatenation. This is the gist behind the `LogicT` transformer (see [_Backtracking, Interleaving, and Terminating Monad Transformers_](https://okmij.org/ftp/papers/LogicT.pdf)).

It can be understood in terms of the single-continuation monad, though I believe:
Take `type ContT2 r m = ContT () (ContT r m)`

Then  the alternative instance becomes, I believe

```haskell
instance Alternative (ContT2 r m) where
  (Cont f) <|> (Cont g) = ContT $ \k -> a k >> bk
```

I'm really thinking aloud here, but this looks a lot like the initial proposal (I suppose we can say that that `r = ()` is a monoid is used here).

---

If we go back to maths, `Alternative f` says “there's a monoid on `f x` for all `x`”, whereas we only need a `Monoid (m x)`, so `Alternative f` is probably a little too strong.

This is what largely I was going for with my `Additive` story in the library. What you probably want (still thinking aloud here) would be

```haskell
instance Additive (m x) => Alternative (ContT r m)
```

Then, you have both versions.
- For instance my instance is obtained by the observation up to some newtype you have that `ContT r m x` is `Additive` if `Monoid x`, because it's coercible to the [`Ap`](https://hackage-content.haskell.org/package/base-4.22.0.0/docs/Data-Monoid.html#t:Ap) monoid.
- Or any applicative as in the original version
- Or any alternative as in the revised version

I guess my conclusion is that these aren't necessarily mathematically competing designs.

I suppose `Alternative` is more natural. And maybe we can exploit the fact that it has almost no law to give an alternative instance to `Const`:

```haskell
instance Monoid a => Alternative (Const a) where
  empty = Const mempty
  (Const l) <> (Const r) = Const (l <> r)
```

(yes this is the same as the applicative instance, but I think nothing forbids us)

So at the price of an almost dishonest amount of cheekiness, the version with  `Alternative` may even be the most general.

---

Sorry… this is too long, I'm not sure it helps.

**Comment by sjshuck** [comment]
att'n @L0neGamer 
At this point there are a few options laid out, and none have been identified as problematic.  The only one suggested with an actual use case is the original one by dfranke.  It's also the one that's actually been requested.  I would be satisfied with it.

**Comment by L0neGamer** [comment]
It feels odd to me for us to have an Alternative that doesn't "short circuit" - with a `liftA2` definition, both actions will always be run.

**Comment by sjshuck** [comment]
That was also Eric's argument.

It depends on the `Applicative m` and `Monoid r` instances.  For OP's use case 1, the monoid short-circuits.
In general I don't think short-circuiting is a requirement of `<|>`.  For instance, list-like structures like `Vector` unconditionally "run" the right-hand side when the asum is forced.

**Comment by L0neGamer** [comment]
Yes, but in a list-like case liftA2 gives the cartesian product instead of a concatenation.

The most obvious case I can think of is some IO actions. With the proposed definition, both IO actions are always run, which is very different from IO's usual `Alternative` which runs each action in left to right order as long as they don't throw.

**Comment by sjshuck** [comment]
Both the `m` and `r` need to be right for short-circuiting to happen in OP's instance.  I agree that it can't be `MonadIO`.  My point in playing devil's advocate was that one need not expect `<|>` to short-circuit.

I can submit a PR with an example plus test to demonstrate use case, which as usual is the thing these discussions tend to be wanting.

**Comment by L0neGamer** [comment]
I'm happy to see an example.
It feels like it'd be confusing that you're expecting "choose the effects of one of these actions" and you get "all the effects of all actions"
